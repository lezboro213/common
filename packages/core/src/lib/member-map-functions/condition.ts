import type {
    ConditionPredicate,
    ConditionReturn,
    Dictionary,
    SelectorReturn,
} from '../types';
import { TransformationType } from '../types';
import { get } from '../utils/get';

export function condition<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    predicate: ConditionPredicate<TSource>,
    defaultValue?: TSelectorReturn
): ConditionReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.Condition,
        (source, sourceMemberPaths) => {
            if (predicate(source)) {
                return get(source, sourceMemberPaths) as TSelectorReturn;
            }

            return defaultValue as TSelectorReturn;
        },
    ];
}
