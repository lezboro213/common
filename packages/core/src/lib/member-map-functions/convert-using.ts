import type {
    Converter,
    ConvertUsingReturn,
    Dictionary,
    Selector,
    SelectorReturn,
} from '../types';
import { TransformationType } from '../types';

export function convertUsing<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>,
    TConvertSourceReturn = SelectorReturn<TSource>
>(
    converter: Converter<TConvertSourceReturn, TSelectorReturn>,
    selector: Selector<TSource, TConvertSourceReturn>
): ConvertUsingReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.ConvertUsing,
        (source) => converter.convert(selector(source)),
    ];
}
