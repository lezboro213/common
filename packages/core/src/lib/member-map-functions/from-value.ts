import type { Dictionary, FromValueReturn, SelectorReturn } from '../types';
import { TransformationType } from '../types';

export function fromValue<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    rawValue: TSelectorReturn
): FromValueReturn<TSource, TDestination, TSelectorReturn> {
    return [TransformationType.FromValue, () => rawValue];
}
