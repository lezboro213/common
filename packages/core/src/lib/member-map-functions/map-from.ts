import type {
    Dictionary,
    MapFromReturn,
    Resolver,
    SelectorReturn,
    ValueSelector,
} from '../types';
import { TransformationType } from '../types';
import { isResolver } from '../utils/is-resolver';

export function mapFrom<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    from:
        | ValueSelector<TSource, TDestination, TSelectorReturn>
        | Resolver<TSource, TDestination, TSelectorReturn>
): MapFromReturn<TSource, TDestination, TSelectorReturn> {
    if (isResolver(from)) {
        return [TransformationType.MapFrom, from.resolve.bind(from)];
    }

    return [TransformationType.MapFrom, from];
}
