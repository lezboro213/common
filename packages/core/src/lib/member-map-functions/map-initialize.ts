import type { Dictionary, MapInitializeReturn, SelectorReturn } from '../types';
import { TransformationType } from '../types';
import { get } from '../utils/get';

export function mapInitialize<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    sourcePath: string[]
): MapInitializeReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.MapInitialize,
        (source) => get(source, sourcePath) as TSelectorReturn,
    ];
}
