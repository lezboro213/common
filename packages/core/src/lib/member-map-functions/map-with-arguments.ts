import type {
    Dictionary,
    MapWithArgumentsReturn,
    Resolver,
    SelectorReturn,
} from '../types';
import { TransformationType } from '../types';
import { isResolver } from '../utils/is-resolver';

export function mapWithArguments<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    withArgumentsResolver:
        | ((
              source: TSource,
              extraArguments: Record<string, unknown>
          ) => TSelectorReturn)
        | Resolver<TSource, Record<string, unknown>, TSelectorReturn>
): MapWithArgumentsReturn<TSource, TDestination, TSelectorReturn> {
    if (isResolver(withArgumentsResolver)) {
        return [
            TransformationType.MapWithArguments,
            withArgumentsResolver.resolve.bind(withArgumentsResolver),
        ];
    }
    return [TransformationType.MapWithArguments, withArgumentsResolver];
}
