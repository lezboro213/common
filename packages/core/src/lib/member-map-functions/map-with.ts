import type {
    Dictionary,
    MapOptions,
    MapWithReturn,
    ModelIdentifier,
    SelectorReturn,
} from '../types';
import { TransformationType, ValueSelector } from '../types';

type Constructor<TModel> = new (...args: unknown[]) => TModel;

export function mapWith<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>,
    TWithDestination extends ModelIdentifier = ModelIdentifier,
    TWithSource extends ModelIdentifier = ModelIdentifier,
    TWithSourceValue extends ValueSelector = TWithSource extends Constructor<
        infer InferredWithSource
    >
        ? ValueSelector<TSource, InferredWithSource>
        : ValueSelector<TSource>
>(
    withDestination: TWithDestination,
    withSource: TWithSource,
    withSourceValue: TWithSourceValue
): MapWithReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.MapWith,
        (source, mapper, options) => {
            const sourceValue = withSourceValue(source);

            if (Array.isArray(sourceValue)) {
                return mapper.mapArray(
                    sourceValue,
                    withSource,
                    withDestination,
                    options as unknown as MapOptions<TSource[], TDestination[]>
                ) as unknown as TSelectorReturn;
            }

            return mapper.map(
                sourceValue,
                withSource,
                withDestination,
                options
            ) as unknown as TSelectorReturn;
        },
    ];
}
