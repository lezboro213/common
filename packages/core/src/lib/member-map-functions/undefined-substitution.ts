import type {
    Dictionary,
    SelectorReturn,
    UndefinedSubstitutionReturn,
} from '../types';
import { TransformationType } from '../types';
import { get } from '../utils/get';

export function undefinedSubstitution<
    TSource extends Dictionary<TSource>,
    TDestination extends Dictionary<TDestination>,
    TSelectorReturn = SelectorReturn<TDestination>
>(
    substitution: TSelectorReturn
): UndefinedSubstitutionReturn<TSource, TDestination, TSelectorReturn> {
    return [
        TransformationType.UndefinedSubstitution,
        (source, sourceMemberPath) => {
            const sourceValue = get(
                source,
                sourceMemberPath
            ) as TSelectorReturn;
            return sourceValue === undefined ? substitution : sourceValue;
        },
    ];
}
