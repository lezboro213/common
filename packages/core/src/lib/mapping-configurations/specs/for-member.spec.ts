import { fromValue } from '../../member-map-functions/from-value';
import type { Mapping, Mapper } from '../../types';
import { MappingClassId } from '../../types';
import { forMember } from '../for-member';

describe(forMember.name, () => {
    it.skip('should update mapping properties with member map fn', () => {
        const mapping = [] as unknown as Mapping;
        mapping[MappingClassId.mapper] = {} as Mapper;
        mapping[MappingClassId.properties] = [];

        const memberMapFn = fromValue('foo');
        forMember((d: { foo: string }) => d.foo, memberMapFn)(mapping);

        expect(mapping[MappingClassId.properties]).toMatchSnapshot();
    });
});
