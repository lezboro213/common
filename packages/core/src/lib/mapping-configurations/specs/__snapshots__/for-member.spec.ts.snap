// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forMember should update mapping properties with member map fn 1`] = `
Array [
  Array [
    Array [
      "foo",
    ],
    Array [
      Array [
        "foo",
      ],
      Array [
        Array [
          3,
          [Function],
        ],
        undefined,
      ],
    ],
  ],
]
`;
